### PR TITLE
add flux_content API

### PR DESCRIPTION
**Pull request by garlick**
This pr adds public API functions for manipulating the content store, sufficient to replace all the places where we were crafting content store RPC's by hand. 

Along the way, `libutil/shastring` was converted to a more abstract `libutil/blobref` interface that supports arbitrary hash functions, sha256 support was added to demonstrate/test this capability, and users that were hardwired to sha1 were updated to use this interface.

Finally, the `flux-content store --dry-run` option was dropped and a test program that performs hashing of blobs without writing them to the content store was added, as this was not a useful user-facing capability.

### CODE REVIEW

**Comment by garlick** [comment]
er just spotted a bug in the way individual blobrefs are computed when a large blob is split.  Fix coming.

**Comment by grondo** [comment]
Cool! Should I be worried that test coverage of the added files `treeobj.c` and `content.c` is low? ;-)

**Comment by garlick** [comment]
Yeah I will fix that.

**Comment by garlick** [comment]
Rebased on current master, fixed a bug in treeobj and fleshed out its unit tests a bit more.  Not done, just a checkpoint before lunch

**Comment by garlick** [comment]
Getting happier with the coverage, now I'll write up a man page for the flux_content interfaces.

**Comment by grondo** [comment]
@garlick, is more work happening here or is this ready for a review and merge?

**Comment by garlick** [comment]
I have some changes queued up that I'd like to tidy up and push, then it will be ready

File: doc/man3/flux_content_load.adoc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+values described below.
+
+`flux_content_store_get()` completes a store operation, blocking on
+response(s) if needed, parsing the result, and returning a blobref that can
+be used later to retrieve the stored blob.  The blobref is valid until
+`flux_content_destroy()` is called.  If the store request was split
+into multiple store requests internally, this function will fail with errno
+set to EFBIG, and `flux_content_store_get_valref()` should be used to
+retrieve the result as a treeref instead instead.  This function is
+idempotent.
```
Comment (grondo) [inline doc/man3/flux_content_load.adoc]: I could be confused, but should "retrieve the result as treeref" here be "retrieve results as valref" instead? I'm basing this solely on the function name "`flux_content_store_get_valref`" as I don't quite understand the API fully yet.

**Comment by garlick** [comment]
No you're right.  Thanks for caching that!  Let me push my changes as I expanded that man page a bit.

**Comment by garlick** [comment]
@grondo verbally suggested that the "content store" should be referred to as the "content service" to avoid confusion since "store" is one of the operations performed by the service.  (This is not to mention the confusing "store_get" and "load_get", where the content service and rpc idioms are mixed!)  Just pushed an update to the man pages to hopefully make this a little better.

**Comment by grondo** [comment]
The only thing that is still slightly confusing is the valref v. blobref case in `flux_content_store_get`. If a valref can refer to *one* or more blobrefs, would it make sense to have the basic function always return a valref? Otherwise, I guess every call to `flux_content_store_get` has to check for `E2BIG` and then call `flux_content_store_get_valref`? My worry would be that most callers would skip that check and have an unexpected failure the one time out of 1000 that the content store request was too large.

I could be worried over nothing though, and since the main user here is the kvs, I think you would know best...

**Comment by garlick** [comment]
I expected that a caller would want either blobrefs or valrefs.  If the former, maxblob would be set to zero, and a blob that is too large for the content service would receive an error.  If the latter, maxblob would be set to some value (perhaps from getattr content-blob-size-limit), and the result  would always be retrieved as a valref.

Besides the misleading man page, I am thinking I have two abstractions mixed up here.  Maybe this interface should simply be for single blobrefs and a different interface should be proposed later that covers all the KVS metadata?

**Comment by grondo** [comment]
> I expected that a caller would want either blobrefs or valrefs. If the former, maxblob would be set to zero, and a blob that is too large for the content service would receive an error. If the latter, maxblob would be set to some value (perhaps from getattr content-blob-size-limit), and the result would always be retrieved as a valref.

Ah! that makes sense! I had assumed (probably due to lazy manpage reading) that a hard max blob size would be configured in the content service -- i.e. a caller to `store` may not know beforehand if the store operation would potentially be split. Am I way off?

> Maybe this interface should simply be for single blobrefs and a different interface should be proposed later that covers all the KVS metadata?

I am not sure -- I think I'd defer to you here. One short term use case that comes to mind, is a module loading system that could store or map a module into the content store and send out a valref to the module loading service.

A near term use case

**Comment by garlick** [comment]
The content service uses the `content-blob-size-limit` broker attribute to set its maximum.  The caller can get this and do the fragmentation on the client side so that it fits with what the server can handle.  The main point of the fragmentation is to avoid problems with really huge messages, head of line blocking and all that, so fragmentation on the client is key.

**Comment by grondo** [comment]
Ok, I'm sorry for not doing my homework! I had assumed the interface would not allow `flux_content_store` with `size` > `content-blob-size-limit`, but in this API that is the responsibility of the caller?  That should get a mention and/or convenience function in the manpage perhaps

**Comment by garlick** [comment]
No it really does feel like this design is inherently confusing.

Let me see if I can make it less so.

**Comment by garlick** [comment]
I scaled the content API back to directly cover the content service operations, and to not grok "valref" at all.   This tidies up the few places where those RPCs were being constructed by hand, and as before, also gets rid of the hardwired sha1 code.

The new API returns a `flux_rpc_t` so it is no longer necessary to have duplicate functions for `flux_rpc_then()` and `flux_rpc_check()` - those just work.  API is now:
```C
/* flags */
enum {
    CONTENT_FLAG_CACHE_BYPASS = 1,/* request direct to backing store */
    CONTENT_FLAG_UPSTREAM = 2,    /* make request of upstream TBON peer */
};

/* Send request to load blob by blobref.
 */
flux_rpc_t *flux_content_load (flux_t *h, const char *blobref, int flags);

/* Get result of load request (blob).
 * This blocks until response is received.
 * Storage for 'buf' belongs to 'r' and is valid until 'r' is destroyed.
 * Returns 0 on success, -1 on failure with errno set.
 */
int flux_content_load_get (flux_rpc_t *rpc, void *buf, int *len);

/* Send request to store blob.
 */
flux_rpc_t *flux_content_store (flux_t *h, const void *buf, int len, int flags);

/* Get result of store request (blobref).
 * Storage for 'blobref' belongs to 'r' and is valid until 'r' is destroyed.
 * Returns 0 on success, -1 on failure with errno set.
 */
int flux_content_store_get (flux_rpc_t *r, const char **blobref);
'''

**Comment by grondo** [comment]
Wow, that is really nice now, kudos! I really like the idiom here of reusing `flux_rpc_then` and `flux_rpc_check`, with custom `flux_*_get`. Nice work!

**Comment by garlick** [comment]
Thanks, well dropping "treeobj" support made it possible to do in the obvious way.  This interface offers no help if you want to store something greater than the configured max blob size, but I think that's probably OK here.

I'll rebase after #909 goes in.

**Comment by garlick** [comment]
OK, I've rebased this.

**Comment by grondo** [comment]
Great! Can I click the button?

**Comment by garlick** [comment]
Please :-)

**Pull request merged**